mautrix_signal/portal.py

```diff
@@ -522,6 +548,71 @@ async def handle_matrix_avatar(self, user: 'u.User', url: ContentURI) -> None:
 
+    @classmethod
+    async def _expire_event(cls, disappearing_message: DisappearingMessage):
+        """
+        Expire a :class:`DisappearingMessage`. This should only be called once the message has been
+        read, as the timer for redaction will start immediately, and there is no (supported)
+        mechanism to stop the countdown, even after bridge restart.
+        """
+        room_id = disappearing_message.room_id
+        event_id = disappearing_message.mxid
+
+        portal = await cls.get_by_mxid(room_id)
+        if not portal:
+            raise AttributeError(f"No portal found for {room_id}")
+
+        if (
+            not portal.is_direct and not
+            cls.config["signal.enable_disappearing_messages_in_groups"]
+        ):
+            portal.log.debug(
+                "Not expiring event in group message since "
+                "signal.enable_disappearing_messages_in_groups is not enabled."
+            )
+            await disappearing_message.delete(room_id, event_id)
+            return
+
+        wait = disappearing_message.expiration_seconds
+        # If there is an expiration_ts, then there was probably a bridge restart, so we have to
+        # resume the countdown. This is fairly likely to occur if the disappearance timeout is
+        # weeks.
+        # If there is not an expiration_ts, then set one.
+        now = time.time()
+        if disappearing_message.expiration_ts is not None:
+            wait = (disappearing_message.expiration_ts / 1000) - now
+        else:
+            disappearing_message.expiration_ts = int((now + wait) * 1000)
+            await disappearing_message.update()
+
+        if wait < 0:
+            wait = 0
+
+        portal.log.debug(f"Redacting {event_id} in {wait} seconds")
+        await asyncio.sleep(wait)
+
+        async with portal._expiration_lock:
+            if not await DisappearingMessage.get(room_id, event_id):
+                portal.log.debug(
+                    f"{event_id} no longer in disappearing messages list, not redacting"
+                )
+                return
+
+            portal.log.debug(f"Redacting {event_id} because it was expired")
+            try:
+                await portal.main_intent.redact(room_id, event_id)
+                portal.log.debug(f"Redacted {event_id} successfully")
+            except Exception as e:
+                portal.log.warning("Redacting expired event didn't work", e)
+            finally:
+                await DisappearingMessage.delete(room_id, event_id)
+
+    async def handle_read_receipt(self, event_id: EventID, data: SingleReceiptEventContent):
+        # Start the redaction timers for all of the disappearing messages in the room when the user
+        # reads the room. This is the behavior of the Signal clients.
+        for disappearing_message in await DisappearingMessage.get_all_for_room(self.mxid):
+            asyncio.create_task(Portal._expire_event(disappearing_message))
```

Comment:
Won't this create a new task for each message in the room whenever there's a read receipt?

Reply:
oh, good point. Let me fix...